File: kotlin-vs-java/src/main/java/SideBySide.java
import java.nio.charset.StandardCharsets;

/**
 * Java Example
 */
public class SideBySide {
  public static void main(String[] args) {
    Car car = new Car("VW", "Passat", 2001);

    System.out.printf("%s, %s\n", car.getMake(), car.model);

    String name = "Ray";
    if (name == "Ray") {
      System.out.println("Your name is Ray!");
    }
  }

  // 1. Semicolons
  public void semicolons() {
    System.out.println("Hello from Java!");
  }

  // 2. Nullability
  public void nullpointers(String s) {
    byte[] bytes = s.getBytes(StandardCharsets.UTF_8);
  }

  // 3. Data Class
  public static class Car {
    private final String make;
    private final String model;

    public String getMake() {
      return make;
    }

    public String getModel() {
      return model;
    }

    public int getYear() {
      return year;
    }

    private final int year;

    public Car(String make, String model, int year) {
      this.make = make;
      this.model = model;
      this.year = year;
    }
  }
}

File: kotlin-vs-java/src/main/kotlin/SideBySide.kt
import kotlinx.coroutines.experimental.async
import java.nio.charset.StandardCharsets

/**
 * Kotlin Example
 *
 * Refer to [Car]
 */
fun main(args: Array<String>) {
    val mycar = Car("VW", "Passat", 2001)

    val yourcar = Car(make = "BMW",
            model = "M3")

    println("${yourcar.make}")

    println("""
Hello there... This
is a
multi-line string!
""")

    val name = "Ray"

    if (name == "Ray") {
        println("Your name is Ray!")
    }

    val c1 = Color(100, 100, 100)
    val c2 = Color(21, 32, 34)
    val c3 = c1 + c2

    println(c3)

    c3.darken()

    async {

    }
}

// 1. Semicolons
fun semicolons() {
    println("Hello from Java!")
}

// 2. Nullability
fun nullpointers(s: String) {
    val bytes = s.toByteArray(StandardCharsets.UTF_8)
}

// 3. Data Class
// 4. Named Parameter
// 5. Default Values
data class Car(val make : String,
               val model : String,
               val year : Int = 2010)

// 6. Setter/Getter on Properties


// 7. String Template
// 8. Multi-line String
// 9. Equality vs. Identity

// 10. Operator Overloading
data class Color(val r : Int,
                 val g : Int,
                 val b : Int) {
    var name : String = ""
    set(value) {
        if (name.isEmpty()) {
            throw IllegalArgumentException();
        }
        this.name = value;
    }
}

operator fun Color.plus(another : Color) : Color {
    return Color (this.r + another.r,
            this.g + another.g,
            this.b + another.b)
}

// 11. Extension Functions
fun Color.darken() : Color {
    return Color(this.r - 1,
            this.g - 1,
            this.b - 1)
}

// 12. KDoc vs JavaDoc!

// Watch out though!
// 1. Public by default!
// 2. Design for override - overriding is explicit! (open, override)
// 3. No ternary operator

open class Vehicle(licensePlate : String, wheels : Int) {
    open fun operate() {
        println("driving the vehicle")
    }
}

class Motorcycle(licensePlate : String, wheels : Int) :
        Vehicle(licensePlate, wheels) {
    override fun operate() {
        println("riding the motorcycle")
    }
}

